lib/srv/db/spanner/grpcserver.go

```diff
@@ -0,0 +1,467 @@
+	ctx := stream.Context()
+	skipAudit := isQuerySpam(req.GetSql())
+	if err := e.authorizeRPCRequest(ctx, req.GetSession(), "ExecuteStreamingSql", req, skipAudit); err != nil {
+		return trace.Wrap(err)
+	}
+	clt, err := e.getClient(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	cc, err := clt.ExecuteStreamingSql(ctx, req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return trace.Wrap(proxyServerStream(cc, stream))
+}
+
+func (e *Engine) StreamingRead(req *spannerpb.ReadRequest, stream spannerpb.Spanner_StreamingReadServer) error {
+	ctx := stream.Context()
+	if err := e.authorizeRPCRequest(ctx, req.GetSession(), "StreamingRead", req, false); err != nil {
+		return trace.Wrap(err)
+	}
+	clt, err := e.getClient(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	cc, err := clt.StreamingRead(ctx, req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return trace.Wrap(proxyServerStream(cc, stream))
+}
+
+func (e *Engine) getClient(ctx context.Context) (spannerpb.SpannerClient, error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	// re-use the connection if we already connected.
+	if e.gcloudClient != nil {
+		return spannerpb.NewSpannerClient(e.gcloudClient), nil
+	}
+
+	tlsCfg, err := e.Auth.GetTLSConfig(ctx, e.sessionCtx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// get a token source for RPC auth. Assume (essentially require) that the
+	// db username is the <name> portion of <name>@<project>.iam.gserviceaccount.com.
+	// TODO(gavin): should we relax the naming requirements? Is there ever a
+	// reason someone would want to use the full service account name, like for
+	// a service account in a different project?
+	dbUser := databaseUserToGCPServiceAccount(e.sessionCtx)
+	ts, err := e.Auth.GetSpannerTokenSource(ctx, e.sessionCtx.WithUser(dbUser))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// The context passed to DialPool controls dialing, but more importantly it
+	// controls GCP oauth2 credential refreshing.
+	// We do not want credential refreshing to stop working after the first
+	// oauth token expires (by default after one hour).
+	// Therefore use the client connection context rather than the RPC context.
+	cc, err := gtransport.DialPool(e.clientConnContext,
+		append(spannerapi.DefaultClientOptions(),
+			option.WithGRPCDialOption(grpc.WithTransportCredentials(credentials.NewTLS(tlsCfg))),
+			option.WithTokenSource(ts),
+			option.WithEndpoint(e.sessionCtx.Database.GetURI()),
+			// pool size seems to be adjusted to number of gRPC channels, which is
+			// 4 by default in Google's code and seemingly other clients as well,
+			// but with Teleport in the middle each downstream client connection
+			// will be handled by a separate engine.
+			// Therefore, avoid amplifying the number of connections dialed to
+			// GCP and just use a pool size of 1.
+			option.WithGRPCConnectionPool(1))...,
```

Comment:
yes, but only 1 connection.

If you configure a client to use 4 gRPC channels, it will make 4 separate connections and each connection will be routed into its own engine, possibly on different database agents if you have multiple agents heartbeating the db.

However a gRPC channel can be used for concurrent requests. So you can make many RPCs at once over a channel and all of them will hit the same db agent engine.